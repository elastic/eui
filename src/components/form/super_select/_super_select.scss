/*
 * 1. Make popover the same width as the form control
 * 2. Style popover similar to combobox
 * 3. Use attribute selector to match popover position without needing the full popover class name
 */

.euiSuperSelect {
  &:not(.euiSuperSelect--fullWidth) { /* 1 */
    // sass-lint:disable-block no-important
    max-width: $euiFormMaxWidth !important; // override default popover styles
  }
}

.euiSuperSelect__popoverPanel[class*='bottom'] { /* 3 */
  border-top-color: transparentize($euiBorderColor, .2);
  border-top-right-radius: 0; /* 2 */
  border-top-left-radius: 0; /* 2 */
}

.euiSuperSelect__popoverPanel[class*='top'] { /* 3 */
  @include euiBottomShadowFlat; /* 2 */

  border-bottom-color: transparentize($euiBorderColor, .2);
  border-bottom-right-radius: 0; /* 2 */
  border-bottom-left-radius: 0; /* 2 */
}

.euiSuperSelect__item {
  &:hover,
  &:focus {
    text-decoration: none;
    background-color: $euiFocusBackgroundColor;
  }
}

.euiSuperSelect__item--hasDividers:not(:last-of-type) {
  border-bottom: $euiBorderThin;
}

.euiSuperSelect__item--compressed {
  padding: $euiSizeS;
  font-size: $euiFontSizeS;
}
