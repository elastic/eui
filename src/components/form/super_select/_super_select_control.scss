/**
 * 1. Leave room for caret.
 * 2. Ensure the descenders don't get cut off
 * 3. Makes sure the height is correct when there's no selection
 */

.euiSuperSelectControl {
  @include euiFormControlStyle;
  @include euiFormControlWithIcon($side: 'right'); /* 1 */
  @include euiFormControlIsLoading($isNextToIcon: true);

  display: block; /* 3 */
  text-align: left;
  line-height: $euiFormControlHeight; /* 2 */
  padding-top: 0; /* 2 */
  padding-bottom: 0; /* 2 */
  // Truncate the text
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;

  &-isInvalid {
    @include euiFormControlInvalidStyle;
  }

  &--compressed {
    @include euiFormControlWithIcon($side: 'right', $compressed: true); /* 1 */
    line-height: $euiFormControlCompressedHeight; /* 2 */
    padding-top: 0; /* 2 */
    padding-bottom: 0; /* 2 */
  }

  &.euiSuperSelect--isOpen__button { // since this is a button, we also want the visual indicator of active when options are shown
    @include euiFormControlFocusStyle;
  }
}
