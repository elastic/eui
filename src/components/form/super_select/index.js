export {
  EuiSuperSelect,
} from './super_select';

export {
  EuiSuperSelectControl,
} from './super_select_control';
