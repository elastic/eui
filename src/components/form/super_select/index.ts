export { EuiSuperSelect, EuiSuperSelectProps } from './super_select';

export {
  EuiSuperSelectControl,
  EuiSuperSelectControlProps,
  EuiSuperSelectOption,
} from './super_select_control';
