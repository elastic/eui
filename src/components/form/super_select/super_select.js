import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { EuiScreenReaderOnly } from '../../accessibility';
import { EuiSuperSelectControl } from './super_select_control';
import { EuiPopover } from '../../popover';
import { EuiContextMenuItem } from '../../context_menu';
import { keyCodes } from '../../../services';
import { EuiI18n } from '../../i18n';

const SHIFT_BACK = 'back';
const SHIFT_FORWARD = 'forward';

export class EuiSuperSelect extends Component {
  constructor(props) {
    super(props);

    this.itemNodes = [];
    this.state = {
      isPopoverOpen: props.isOpen || false,
      menuWidth: null,
    };
  }

  componentDidMount() {
    this._isMounted = true;
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  setItemNode = (node, index) => {
    this.itemNodes[index] = node;
  };

  setPopoverRef = ref => {
    this.popoverRef = ref;
  };

  openPopover = () => {
    this.setState({
      isPopoverOpen: true,
    });

    const focusSelected = () => {
      const indexOfSelected = this.props.options.reduce(
        (indexOfSelected, option, index) => {
          if (indexOfSelected != null) return indexOfSelected;
          if (option == null) return null;
          return option.value === this.props.valueOfSelected ? index : null;
        },
        null
      );

      requestAnimationFrame(() => {
        if (!this._isMounted) {
          return;
        }
        this.setState({
          menuWidth: this.popoverRef.getBoundingClientRect().width - 2, // account for border not inner shadow
        });

        if (this.props.valueOfSelected != null) {
          if (indexOfSelected != null) {
            this.focusItemAt(indexOfSelected);
          } else {
            focusSelected();
          }
        }
      });
    };

    requestAnimationFrame(focusSelected);
  };

  closePopover = () => {
    this.setState({
      isPopoverOpen: false,
    });
  };

  itemClicked = value => {
    this.setState({
      isPopoverOpen: false,
    });
    this.props.onChange(value);
  };

  onSelectKeyDown = e => {
    if (e.keyCode === keyCodes.UP || e.keyCode === keyCodes.DOWN) {
      e.preventDefault();
      e.stopPropagation();
      this.openPopover();
    }
  };

  onItemKeyDown = e => {
    switch (e.keyCode) {
      case keyCodes.ESCAPE:
        // close the popover and prevent ancestors from handling
        e.preventDefault();
        e.stopPropagation();
        this.closePopover();
        break;

      case keyCodes.TAB:
        // no-op
        e.preventDefault();
        e.stopPropagation();
        break;

      case keyCodes.UP:
        e.preventDefault();
        e.stopPropagation();
        this.shiftFocus(SHIFT_BACK);
        break;

      case keyCodes.DOWN:
        e.preventDefault();
        e.stopPropagation();
        this.shiftFocus(SHIFT_FORWARD);
        break;
    }
  };

  focusItemAt(index) {
    const targetElement = this.itemNodes[index];
    if (targetElement != null) {
      targetElement.focus();
    }
  }

  shiftFocus(direction) {
    const currentIndex = this.itemNodes.indexOf(document.activeElement);
    let targetElementIndex;

    if (currentIndex === -1) {
      // somehow the select options has lost focus
      targetElementIndex = 0;
    } else {
      if (direction === SHIFT_BACK) {
        targetElementIndex =
          currentIndex === 0 ? this.itemNodes.length - 1 : currentIndex - 1;
      } else {
        targetElementIndex =
          currentIndex === this.itemNodes.length - 1 ? 0 : currentIndex + 1;
      }
    }

    this.focusItemAt(targetElementIndex);
  }

  render() {
    const {
      className,
      options,
      valueOfSelected,
      onChange,
      isOpen,
      isInvalid,
      hasDividers,
      itemClassName,
      itemLayoutAlign,
      fullWidth,
      popoverClassName,
      compressed,
      ...rest
    } = this.props;

    const popoverClasses = classNames(
      'euiSuperSelect',
      {
        'euiSuperSelect--fullWidth': fullWidth,
      },
      popoverClassName
    );

    const popoverPanelClasses = classNames('euiSuperSelect__popoverPanel', {
      [`${popoverClassName}__popoverPanel`]: !!popoverClassName,
    });

    const buttonClasses = classNames(
      {
        'euiSuperSelect--isOpen__button': this.state.isPopoverOpen,
      },
      className
    );

    const itemClasses = classNames(
      'euiSuperSelect__item',
      {
        'euiSuperSelect__item--hasDividers': hasDividers,
        'euiSuperSelect__item--compressed': compressed,
      },
      itemClassName
    );

    const button = (
      <EuiSuperSelectControl
        options={options}
        value={valueOfSelected}
        onClick={
          this.state.isPopoverOpen ? this.closePopover : this.openPopover
        }
        onKeyDown={this.onSelectKeyDown}
        className={buttonClasses}
        fullWidth={fullWidth}
        isInvalid={isInvalid}
        compressed={compressed}
        {...rest}
      />
    );

    const items = options.map((option, index) => {
      const { value, dropdownDisplay, inputDisplay, ...optionRest } = option;

      return (
        <EuiContextMenuItem
          key={index}
          className={itemClasses}
          icon={valueOfSelected === value ? 'check' : 'empty'}
          onClick={() => this.itemClicked(value)}
          onKeyDown={this.onItemKeyDown}
          layoutAlign={itemLayoutAlign}
          buttonRef={node => this.setItemNode(node, index)}
          style={{ width: this.state.menuWidth }}
          role="option"
          id={value}
          aria-selected={valueOfSelected === value}
          {...optionRest}>
          {dropdownDisplay || inputDisplay}
        </EuiContextMenuItem>
      );
    });

    return (
      <EuiPopover
        className={popoverClasses}
        display="block"
        panelClassName={popoverPanelClasses}
        button={button}
        isOpen={isOpen || this.state.isPopoverOpen}
        closePopover={this.closePopover}
        panelPaddingSize="none"
        anchorPosition="downCenter"
        ownFocus={false}
        popoverRef={this.setPopoverRef}
        hasArrow={false}>
        <EuiScreenReaderOnly>
          <p role="alert">
            <EuiI18n
              token="euiSuperSelect.screenReaderAnnouncement"
              default="You are in a form selector of {optionsCount} items and must select a single option.
              Use the up and down keys to navigate or escape to close."
              values={{ optionsCount: options.length }}
            />
          </p>
        </EuiScreenReaderOnly>
        <div
          role="listbox"
          aria-activedescendant={valueOfSelected}
          tabIndex="0">
          {items}
        </div>
      </EuiPopover>
    );
  }
}

EuiSuperSelect.propTypes = {
  /**
   * Classes (and `...rest`) will be applied to the control
   */
  className: PropTypes.string,
  /**
   * Classes for the context menu item
   */
  itemClassName: PropTypes.string,
  /**
   * You must pass an `onChange` function to handle the update of the value
   */
  onChange: PropTypes.func,
  /**
   * Pass an array of options that must at least include:
   * `value`: storing unique value of item,
   * `inputDisplay`: what shows inside the form input when selected
   * `dropdownDisplay` (optional): what shows for the item in the dropdown
   */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
      inputDisplay: PropTypes.node,
      dropdownDisplay: PropTypes.node,
    })
  ).isRequired,
  valueOfSelected: PropTypes.string,
  /**
   * Change to `true` if you want horizontal lines between options.
   * This is best used when options are multi-line.
   */
  hasDividers: PropTypes.bool,
  /**
   * Change `EuiContextMenuItem` layout position of icon
   */
  itemLayoutAlign: PropTypes.string,
  /**
   * Make it wide
   */
  fullWidth: PropTypes.bool,
  /**
   * Provides invalid styling
   */
  isInvalid: PropTypes.bool,
  /**
   * Provides loading indictor
   */
  isLoading: PropTypes.bool,
  /**
   * Make it short
   */
  compressed: PropTypes.bool,
  /**
   * Applied to the outermost wrapper (popover)
   */
  popoverClassName: PropTypes.string,
};

EuiSuperSelect.defaultProps = {
  hasDividers: false,
  fullWidth: false,
  compressed: false,
  isInvalid: false,
};
