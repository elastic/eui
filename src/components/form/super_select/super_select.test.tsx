/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { mount, render } from 'enzyme';
import { requiredProps, takeMountedSnapshot } from '../../../test';

import { EuiSuperSelect } from './super_select';

jest.mock('../../portal', () => ({
  EuiPortal: ({ children }: any) => children,
}));

const options = [
  { value: '1', inputDisplay: 'Option #1' },
  { value: '2', inputDisplay: 'Option #2' },
];

describe('EuiSuperSelect', () => {
  test('is rendered', () => {
    const component = render(
      <EuiSuperSelect
        {...requiredProps}
        options={options}
        onChange={() => {}}
      />
    );

    expect(component).toMatchSnapshot();
  });

  describe('props', () => {
    test('fullWidth is rendered', () => {
      const component = render(
        <EuiSuperSelect
          {...requiredProps}
          options={options}
          onChange={() => {}}
          fullWidth
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('compressed is rendered', () => {
      const component = render(
        <EuiSuperSelect
          {...requiredProps}
          options={options}
          onChange={() => {}}
          compressed
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('is rendered with a prepend and append', () => {
      const component = render(
        <EuiSuperSelect
          {...requiredProps}
          options={options}
          onChange={() => {}}
          prepend="prepend"
          append="append"
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('select component is rendered', () => {
      const component = render(
        <EuiSuperSelect
          options={[
            { value: '1', inputDisplay: 'Option #1' },
            { value: '2', inputDisplay: 'Option #2' },
          ]}
          onChange={() => {}}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('options are rendered when select is open', () => {
      const component = mount(
        <EuiSuperSelect
          options={options}
          onChange={() => {}}
          data-test-subj="superSelect"
        />
      );

      component.find('button[data-test-subj="superSelect"]').simulate('click');

      expect(takeMountedSnapshot(component)).toMatchSnapshot();
    });

    test('valueSelected is rendered', () => {
      const component = render(
        <EuiSuperSelect
          options={options}
          valueOfSelected="2"
          onChange={() => {}}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('custom display is propagated to dropdown', () => {
      const component = mount(
        <EuiSuperSelect
          options={[
            {
              value: '1',
              inputDisplay: 'Option #1',
              dropdownDisplay: 'Custom Display #1',
            },
            {
              value: '2',
              inputDisplay: 'Option #2',
              dropdownDisplay: 'Custom Display #2',
            },
          ]}
          onChange={() => {}}
          data-test-subj="superSelect"
        />
      );

      component.find('button[data-test-subj="superSelect"]').simulate('click');

      expect(takeMountedSnapshot(component)).toMatchSnapshot();
    });

    test('more props are propogated to each option', () => {
      const component = mount(
        <EuiSuperSelect
          options={[
            { value: '1', inputDisplay: 'Option #1', disabled: true },
            {
              value: '2',
              inputDisplay: 'Option #2',
              'data-test-subj': 'option two',
            },
          ]}
          valueOfSelected="1"
          onChange={() => {}}
          data-test-subj="superSelect"
        />
      );

      component.find('button[data-test-subj="superSelect"]').simulate('click');

      expect(takeMountedSnapshot(component)).toMatchSnapshot();
    });
  });
});
