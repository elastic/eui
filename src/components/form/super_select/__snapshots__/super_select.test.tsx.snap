// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiSuperSelect is rendered 1`] = `
<div
  class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
>
  <div
    class="euiPopover__anchor"
  >
    <div>
      <input
        type="hidden"
        value=""
      />
      <div
        class="euiFormControlLayout"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <span
            class="euiScreenReaderOnly"
            id="generated-id"
          >
            Select an option: , is selected
          </span>
          <button
            aria-haspopup="true"
            aria-label="aria-label"
            aria-labelledby="undefined generated-id"
            class="euiSuperSelectControl testClass1 testClass2"
            data-test-subj="test subject string"
            type="button"
          />
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <span
                aria-hidden="true"
                class="euiFormControlLayoutCustomIcon__icon"
                data-euiicon-type="arrowDown"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiSuperSelect props compressed is rendered 1`] = `
<div
  class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
>
  <div
    class="euiPopover__anchor"
  >
    <div>
      <input
        type="hidden"
        value=""
      />
      <div
        class="euiFormControlLayout euiFormControlLayout--compressed"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <span
            class="euiScreenReaderOnly"
            id="generated-id"
          >
            Select an option: , is selected
          </span>
          <button
            aria-haspopup="true"
            aria-label="aria-label"
            aria-labelledby="undefined generated-id"
            class="euiSuperSelectControl euiSuperSelectControl--compressed testClass1 testClass2"
            data-test-subj="test subject string"
            type="button"
          />
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <span
                aria-hidden="true"
                class="euiFormControlLayoutCustomIcon__icon"
                data-euiicon-type="arrowDown"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiSuperSelect props custom display is propagated to dropdown 1`] = `
<div
  class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
>
  <div
    class="euiPopover__anchor"
  >
    <div>
      <input
        type="hidden"
        value=""
      />
      <div
        class="euiFormControlLayout"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <span
            class="euiScreenReaderOnly"
            id="generated-id"
          >
            Select an option: , is selected
          </span>
          <button
            aria-haspopup="true"
            aria-labelledby="undefined generated-id"
            class="euiSuperSelectControl euiSuperSelect--isOpen__button"
            data-test-subj="superSelect"
            type="button"
          />
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <span
                aria-hidden="true"
                class="euiFormControlLayoutCustomIcon__icon"
                data-euiicon-type="arrowDown"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
  <div
    data-focus-lock-disabled="disabled"
  >
    <div
      aria-live="assertive"
      aria-modal="true"
      class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--noShadow euiPopover__panel euiPopover__panel--bottom euiPopover__panel-noArrow euiPopover__panel-isAttached"
      role="dialog"
      style="top: 8px; left: 0px; will-change: transform, opacity; z-index: 2000;"
    >
      <div
        class="euiPopover__panelArrow euiPopover__panelArrow--bottom"
      />
      <div>
        <div
          data-focus-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="0"
        />
        <div
          data-focus-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="1"
        />
        <div
          data-focus-lock-disabled="false"
        >
          <div>
            <p
              class="euiScreenReaderOnly"
              role="alert"
            >
              You are in a form selector of 2 items and must select a single option. Use the up and down keys to navigate or escape to close.
            </p>
            <div
              class="euiSuperSelect__listbox"
              role="listbox"
              tabindex="0"
            >
              <button
                aria-selected="false"
                class="euiContextMenuItem euiSuperSelect__item"
                id="1"
                role="option"
                type="button"
              >
                <span
                  class="euiContextMenu__itemLayout"
                >
                  <span
                    class="euiContextMenu__icon"
                    data-euiicon-type="empty"
                  />
                  <span
                    class="euiContextMenuItem__text"
                  >
                    Custom Display #1
                  </span>
                </span>
              </button>
              <button
                aria-selected="false"
                class="euiContextMenuItem euiSuperSelect__item"
                id="2"
                role="option"
                type="button"
              >
                <span
                  class="euiContextMenu__itemLayout"
                >
                  <span
                    class="euiContextMenu__icon"
                    data-euiicon-type="empty"
                  />
                  <span
                    class="euiContextMenuItem__text"
                  >
                    Custom Display #2
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          data-focus-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="0"
        />
      </div>
    </div>
  </div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
</div>
`;

exports[`EuiSuperSelect props fullWidth is rendered 1`] = `
<div
  class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiInputPopover--fullWidth euiSuperSelect"
>
  <div
    class="euiPopover__anchor"
  >
    <div>
      <input
        type="hidden"
        value=""
      />
      <div
        class="euiFormControlLayout euiFormControlLayout--fullWidth"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <span
            class="euiScreenReaderOnly"
            id="generated-id"
          >
            Select an option: , is selected
          </span>
          <button
            aria-haspopup="true"
            aria-label="aria-label"
            aria-labelledby="undefined generated-id"
            class="euiSuperSelectControl euiSuperSelectControl--fullWidth testClass1 testClass2"
            data-test-subj="test subject string"
            type="button"
          />
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <span
                aria-hidden="true"
                class="euiFormControlLayoutCustomIcon__icon"
                data-euiicon-type="arrowDown"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiSuperSelect props is rendered with a prepend and append 1`] = `
<div
  class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
>
  <div
    class="euiPopover__anchor"
  >
    <div>
      <input
        type="hidden"
        value=""
      />
      <div
        class="euiFormControlLayout euiFormControlLayout--group"
      >
        <label
          class="euiFormLabel euiFormControlLayout__prepend"
        >
          prepend
        </label>
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <span
            class="euiScreenReaderOnly"
            id="generated-id"
          >
            Select an option: , is selected
          </span>
          <button
            aria-haspopup="true"
            aria-label="aria-label"
            aria-labelledby="undefined generated-id"
            class="euiSuperSelectControl euiSuperSelectControl--inGroup testClass1 testClass2"
            data-test-subj="test subject string"
            type="button"
          />
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <span
                aria-hidden="true"
                class="euiFormControlLayoutCustomIcon__icon"
                data-euiicon-type="arrowDown"
              />
            </span>
          </div>
        </div>
        <label
          class="euiFormLabel euiFormControlLayout__append"
        >
          append
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiSuperSelect props more props are propogated to each option 1`] = `
<div
  class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
>
  <div
    class="euiPopover__anchor"
  >
    <div>
      <input
        type="hidden"
        value="1"
      />
      <div
        class="euiFormControlLayout"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <span
            class="euiScreenReaderOnly"
            id="generated-id"
          >
            Select an option: Option #1, is selected
          </span>
          <button
            aria-haspopup="true"
            aria-labelledby="undefined generated-id"
            class="euiSuperSelectControl euiSuperSelect--isOpen__button"
            data-test-subj="superSelect"
            type="button"
          >
            Option #1
          </button>
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <span
                aria-hidden="true"
                class="euiFormControlLayoutCustomIcon__icon"
                data-euiicon-type="arrowDown"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
  <div
    data-focus-lock-disabled="disabled"
  >
    <div
      aria-live="assertive"
      aria-modal="true"
      class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--noShadow euiPopover__panel euiPopover__panel--bottom euiPopover__panel-noArrow euiPopover__panel-isAttached"
      role="dialog"
      style="top: 8px; left: 0px; will-change: transform, opacity; z-index: 2000;"
    >
      <div
        class="euiPopover__panelArrow euiPopover__panelArrow--bottom"
      />
      <div>
        <div
          data-focus-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="0"
        />
        <div
          data-focus-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="1"
        />
        <div
          data-focus-lock-disabled="false"
        >
          <div>
            <p
              class="euiScreenReaderOnly"
              role="alert"
            >
              You are in a form selector of 2 items and must select a single option. Use the up and down keys to navigate or escape to close.
            </p>
            <div
              aria-activedescendant="1"
              class="euiSuperSelect__listbox"
              role="listbox"
              tabindex="0"
            >
              <button
                aria-selected="true"
                class="euiContextMenuItem euiSuperSelect__item euiContextMenuItem-isDisabled"
                disabled=""
                id="1"
                role="option"
                type="button"
              >
                <span
                  class="euiContextMenu__itemLayout"
                >
                  <span
                    class="euiContextMenu__icon"
                    data-euiicon-type="check"
                  />
                  <span
                    class="euiContextMenuItem__text"
                  >
                    Option #1
                  </span>
                </span>
              </button>
              <button
                aria-selected="false"
                class="euiContextMenuItem euiSuperSelect__item"
                data-test-subj="option two"
                id="2"
                role="option"
                type="button"
              >
                <span
                  class="euiContextMenu__itemLayout"
                >
                  <span
                    class="euiContextMenu__icon"
                    data-euiicon-type="empty"
                  />
                  <span
                    class="euiContextMenuItem__text"
                  >
                    Option #2
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          data-focus-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="0"
        />
      </div>
    </div>
  </div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
</div>
`;

exports[`EuiSuperSelect props options are rendered when select is open 1`] = `
<div
  class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
>
  <div
    class="euiPopover__anchor"
  >
    <div>
      <input
        type="hidden"
        value=""
      />
      <div
        class="euiFormControlLayout"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <span
            class="euiScreenReaderOnly"
            id="generated-id"
          >
            Select an option: , is selected
          </span>
          <button
            aria-haspopup="true"
            aria-labelledby="undefined generated-id"
            class="euiSuperSelectControl euiSuperSelect--isOpen__button"
            data-test-subj="superSelect"
            type="button"
          />
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <span
                aria-hidden="true"
                class="euiFormControlLayoutCustomIcon__icon"
                data-euiicon-type="arrowDown"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
  <div
    data-focus-lock-disabled="disabled"
  >
    <div
      aria-live="assertive"
      aria-modal="true"
      class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--noShadow euiPopover__panel euiPopover__panel--bottom euiPopover__panel-noArrow euiPopover__panel-isAttached"
      role="dialog"
      style="top: 8px; left: 0px; will-change: transform, opacity; z-index: 2000;"
    >
      <div
        class="euiPopover__panelArrow euiPopover__panelArrow--bottom"
      />
      <div>
        <div
          data-focus-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="0"
        />
        <div
          data-focus-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="1"
        />
        <div
          data-focus-lock-disabled="false"
        >
          <div>
            <p
              class="euiScreenReaderOnly"
              role="alert"
            >
              You are in a form selector of 2 items and must select a single option. Use the up and down keys to navigate or escape to close.
            </p>
            <div
              class="euiSuperSelect__listbox"
              role="listbox"
              tabindex="0"
            >
              <button
                aria-selected="false"
                class="euiContextMenuItem euiSuperSelect__item"
                id="1"
                role="option"
                type="button"
              >
                <span
                  class="euiContextMenu__itemLayout"
                >
                  <span
                    class="euiContextMenu__icon"
                    data-euiicon-type="empty"
                  />
                  <span
                    class="euiContextMenuItem__text"
                  >
                    Option #1
                  </span>
                </span>
              </button>
              <button
                aria-selected="false"
                class="euiContextMenuItem euiSuperSelect__item"
                id="2"
                role="option"
                type="button"
              >
                <span
                  class="euiContextMenu__itemLayout"
                >
                  <span
                    class="euiContextMenu__icon"
                    data-euiicon-type="empty"
                  />
                  <span
                    class="euiContextMenuItem__text"
                  >
                    Option #2
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          data-focus-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="0"
        />
      </div>
    </div>
  </div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="-1"
  />
</div>
`;

exports[`EuiSuperSelect props select component is rendered 1`] = `
<div
  class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
>
  <div
    class="euiPopover__anchor"
  >
    <div>
      <input
        type="hidden"
        value=""
      />
      <div
        class="euiFormControlLayout"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <span
            class="euiScreenReaderOnly"
            id="generated-id"
          >
            Select an option: , is selected
          </span>
          <button
            aria-haspopup="true"
            aria-labelledby="undefined generated-id"
            class="euiSuperSelectControl"
            type="button"
          />
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <span
                aria-hidden="true"
                class="euiFormControlLayoutCustomIcon__icon"
                data-euiicon-type="arrowDown"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiSuperSelect props valueSelected is rendered 1`] = `
<div
  class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
>
  <div
    class="euiPopover__anchor"
  >
    <div>
      <input
        type="hidden"
        value="2"
      />
      <div
        class="euiFormControlLayout"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <span
            class="euiScreenReaderOnly"
            id="generated-id"
          >
            Select an option: Option #2, is selected
          </span>
          <button
            aria-haspopup="true"
            aria-labelledby="undefined generated-id"
            class="euiSuperSelectControl"
            type="button"
          >
            Option #2
          </button>
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <span
                aria-hidden="true"
                class="euiFormControlLayoutCustomIcon__icon"
                data-euiicon-type="arrowDown"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
